[DOCS-8760] Create charts with RUM custom metrics guide (#37735)
DataDog/documentation: content/en/real_user_monitoring/guide/create-charts-with-rum-custom-metrics.md

---
title: Create Charts with RUM Custom Metrics
description: "Learn how to visualize your RUM custom metrics as charts in dashboards and notebooks."
further_reading:
- link: '/real_user_monitoring/platform/generate_metrics'
  tag: 'Documentation'
  text: 'Generate Custom Metrics from RUM Events'
- link: '/real_user_monitoring/guide/alerting-with-rum'
  tag: 'Documentation'
  text: 'Create alerts with RUM data'
- link: '/real_user_monitoring/platform/dashboards/'
  tag: 'Documentation'
  text: 'RUM Dashboards'
- link: '/dashboards/'
  tag: 'Documentation'
  text: 'Dashboards'
---

## Overview

[RUM-based custom metrics][1] let you summarize event data from your browser and mobile applications and visualize trends for up to 15 months. After you [create a custom metric][1], you can use it anywhere Datadog metrics appear, including dashboards, notebooks, and SLOs.

## Add a RUM custom metric to a dashboard

> 1. Open an existing dashboard or [create one][2].

Reviewer: Link text should be descriptive. "create one" is vague out of context.

```suggestion
1. Open an existing dashboard or [create a dashboard][2].
```

2. Click **Add Widgets** and select a widget type. See [Choose the right widget](#choose-the-right-widget) for guidance.
3. In the widget editor, set the data source to **Metrics**.
4. Enter your custom metric name, for example `rum.sessions.count_by_geography`.
5. Apply tag filters to scope the data, for example `env:prod`.
6. Select a **group by** tag to break down the data, for example `@application.name`.
7. Click **Save** to add the widget to your dashboard.

## Choose the right widget

The following table maps common RUM use cases to widget types:

| Use case | Recommended widget |
|---|---|

> | Error rate or session count over time | timeseries |

Reviewer: `timeseries` is lowercase here but all other widget names in this table are capitalized (`Top List`, `Query Value`, `Distribution`, `Geomap`). Should this be `Timeseries` to match? Same question applies to line 77 where it appears as `**timeseries** widget` in the charting examples section.

> | Error rate or session count over time | timeseries |

Author: ```suggestion
| Error rate or session count over time | Timeseries |
```

| Top pages by load time | Top List |
| Single KPI such as crash-free rate | Query Value |

> | Distribution of load times (requires [percentile aggregation][1]) | Distribution |

Reviewer: The `[1]` reference is doing a lot of work — it’s used for “RUM-based custom metrics,” “create a custom metric,” and “percentile aggregation,” all pointing to the same page. For the “percentile aggregation” table entry, linking to the page root will send readers to the top of a long page. If `generate_metrics.md` has a heading anchor for percentile aggregation, consider linking to it directly (e.g., `/real_user_monitoring/platform/generate_metrics/#percentile-aggregation`) so readers land at the relevant content.

Author: thanks for this helpful feedback! created a new section that the link now goes to.

| Breakdown by country or version | Geomap or Top List |

## Combine metrics with formulas

You can combine multiple RUM custom metrics in a single widget using formulas. This is useful for calculating derived values such as rates and ratios.

For example, to calculate a crash-free rate, create two custom metrics:

> - `rum.sessions.crash_free`—a count of sessions that complete without a crash

Reviewer: Definition-style list items should use `**Term**: Description.` format rather than an em dash. Also capitalize the first word of each description and add a closing period since they’re full sentences.

```suggestion
- **`rum.sessions.crash_free`**: A count of sessions that complete without a crash.
- **`rum.sessions.total`**: A count of all sessions.
```

- `rum.sessions.total`—a count of all sessions

In the widget editor, add both metrics as separate queries, then enter the formula `a / b * 100` to calculate the percentage.

> {{< img src="real_user_monitoring/guide/create-charts-with-custom-metrics/custom-metrics-formulas-example.png" alt="Widget editor showing two RUM custom metric queries combined with a formula to calculate crash-free rate as a percentage" style="width:100%;" >}}

Reviewer: Alt text must end with a period.

```suggestion
{{< img src="real_user_monitoring/guide/create-charts-with-custom-metrics/custom-metrics-formulas-example.png" alt="Widget editor showing two RUM custom metric queries combined with a formula to calculate crash-free rate as a percentage." style="width:100%;" >}}
```

## Use template variables for reusable charts

Template variables let you create dashboards that work across multiple applications, environments, or versions without duplicating widgets.

To add a template variable to a dashboard:

1. Click **Add Variable** at the top of the dashboard.
2. Select **Metrics** as the data source.
3. Choose a tag key such as `application.name`, `env`, or `version`.

After you add the variable, reference it in widget queries using the `$variable_name` syntax. For example, filter a widget to `application.name:$application` to let viewers select which application to display.

## Charting examples

### Track conversion funnel drop-off by country

Use a **Geomap** widget with a custom metric that counts completed checkout sessions grouped by country. Filter to `@session.type:user` to exclude synthetic traffic, and group by the `@geo.country` tag to visualize where users drop off in your conversion funnel.

### Monitor crash-free rate by release version

> Use a **timeseries** widget with the formula pattern described above to track crash-free rate over time. Group by `version` to compare stability across releases and detect regressions introduced by specific builds.

Reviewer: Same capitalization question as the table — should this be `Timeseries` to match the other widget names?

```suggestion
Use a **Timeseries** widget with the formula pattern described above to track crash-free rate over time. Group by `version` to compare stability across releases and detect regressions introduced by specific builds.
```

3. In the widget editor, set the data source to **Metrics**.
4. Enter your custom metric name, for example `rum.sessions.count_by_geography`.
5. Apply tag filters to scope the data, for example `env:prod`.
6. Select a **group by** tag to break down the data, for example `@application.name`.
7. Click **Save** to add the widget to your dashboard.

## Choose the right widget

The following table maps common RUM use cases to widget types:

| Use case | Recommended widget |
|---|---|
| Error rate or session count over time | Timeseries |
| Top pages by load time | Top List |
| Single KPI such as crash-free rate | Query Value |
| Distribution of load times (requires [percentile aggregation][3]) | Distribution |
| Breakdown by country or version | Geomap or Top List |

## Combine metrics with formulas

You can combine multiple RUM custom metrics in a single widget using formulas. This is useful for calculating derived values such as rates and ratios.

For example, to calculate a crash-free rate, create two custom metrics:
- **`rum.sessions.crash_free`**: A count of sessions that complete without a crash.
- **`rum.sessions.total`**: A count of all sessions.

> - `rum.sessions.total`—a count of all sessions

Reviewer 2: ```suggestion
```

## Use template variables for reusable charts

Template variables let you create dashboards that work across multiple applications, environments, or versions without duplicating widgets.

To add a template variable to a dashboard:

1. Click **Add Variable** at the top of the dashboard.
2. Select **Metrics** as the data source.
3. Choose a tag key such as `application.name`, `env`, or `version`.

After you add the variable, reference it in widget queries using the `$variable_name` syntax. For example, filter a widget to `application.name:$application` to let viewers select which application to display.

## Charting examples

### Track conversion funnel drop-off by country

Use a **Geomap** widget with a custom metric that counts completed checkout sessions grouped by country. Filter to `@session.type:user` to exclude synthetic traffic, and group by the `@geo.country` tag to visualize where users drop off in your conversion funnel.

### Monitor crash-free rate by release version

> Use a **Timeseries** widget with the formula pattern described above to track crash-free rate over time. Group by `version` to compare stability across releases and detect regressions introduced by specific builds.

Reviewer 2: ```suggestion
Use a **Timeseries** widget with the formula pattern from the [Combine metrics with formulas](#combine-metrics-with-formulas) section to track crash-free rate over time. Group by `version` to compare stability across releases and detect regressions introduced by specific builds.
```
